### PR TITLE
Treat methods with empty bodies in Protocols as abstract

### LINKED ISSUE #8005: Extending a Protocol leads to unsafe behavior
* Are you reporting a bug, or opening a feature request?
bug

* Please insert below the code you are checking with mypy, or a mock-up repro if the source is private. We would appreciate if you try to simplify your case to a minimal repro.
```
from typing import Protocol

class P(Protocol):
    def foo(self) -> int: ...

class A(P):  # mypy does not flag this
    pass

class B:
    pass

test: P = A()  # mypy does not flag this
test = B()  # mypy flags this

print(A().foo() + 5)  # both error at runtime
print(B().foo() + 5)
```

* What is the actual behavior/output?
`A().foo() + 5` crashes at runtime with `TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'`
`B` is correctly type checked and rejected.

* What is the behavior/output you expect?
According to https://mypy.readthedocs.io/en/latest/protocols.html#defining-subprotocols-and-subclassing-protocols, "Explicitly including a protocol as a base class ... forces mypy to verify that your class implementation is actually compatible with the protocol."
I would expect extending a Protocol to be similar to implementing an interface in Java.
In this particular example, mypy should ideally detect that `A` does not fill in the stub method from `P`, like it does for `B`. Alternatively, if that is not possible, perhaps extending a protocol should not be allowed.

* What are the versions of mypy and Python you are using? Do you see the same issue after installing mypy from Git master?
Python 3.8.0
mypy 0.750+dev.d9dea5f3ad30bb449e4167b3e8476684ded0482e

* What are the mypy flags you are using? (For example --strict-optional)
`mypy --strict test.py`

### LINKED ISSUE #8926: Should concrete implementations be required to implement protocol?
* Are you reporting a bug, or opening a feature request?

Bug

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.

```python

from typing import Protocol


class Person(Protocol):
    first_name: str

    def greeting(self, message: str) -> str:
        ...


class FrenchPerson(Person):
    prenom: str

    def __init__(self, name: str):
        self.prenom = name

    def salutation(self, message: str) -> str:
        return f"{message} {self.prenom}"


def print_person(person: Person) -> str:
    return person.greeting(message="Je m'appelle")


def main():
    fp = FrenchPerson(name='Henri')
    response = print_person(fp)
    print(response)
```

* What is the actual behavior/output?

PEP 544 says you can state that an implementation supports a protocol by
[subclassing the protocol](https://www.python.org/dev/peps/pep-0544/#explicitly-declaring-implementation).

It also says: "If one omits Protocol in the base class list, this would
be a regular (non-protocol) class that must implement Sized."

* What is the behavior/output you expect?

I would hope `mypy` would, either at declaration time or in the `print_person` function, spot that something claimed to be a `Person` but had neither the protocol variable `name` nor the method `greeting`.

It's also a shame that `FrenchPerson` got the `greeting` method added to it from the subclassing.

* What are the versions of mypy and Python you are using?

`mypy` 0.770 and Python 3.8.1

* What are the mypy flags you are using? (For example --strict-optional)

None.

### PR DESCRIPTION
**Pull request by thomkeh**
### Description

closes #8005
closes #8926

Based on the _quick fix_ idea [from this comment](https://github.com/python/mypy/issues/8005#issuecomment-558341465) by @ilevkivskyi . 

For the problem description see #8005. The underlying problem is arguably #2350, but that seems not so easy to fix (there was an attempt in #8111 but it was abandoned). For this PR, I borrowed a lot from #8111 but kept the scope much less ambitious (for example, #8111 also warns on `super()` calls to empty functions, which this PR doesn't).  However, as [pointed out here](https://github.com/python/mypy/issues/2350#issuecomment-256701609), empty bodies are actually a valid implementation if the function may return `None`, so I added that as an additional check. But if that is deemed too confusing, I'm happy to remove that check.

So, in summary, what this PR does: Methods in Protocols are considered abstract if they
1. have an empty function body
2. have a return type that is not compatible with `None` and
3. are not in a stub file.

## Test Plan

Added tests for all the new behavior.

### CODE REVIEW

**Comment by nickgaya** [comment]
Issues #8005 and #8926 are both based on an incorrect user belief that an empty method in a Protocol is abstract.  However this does not match Python's runtime behavior.

Instead of treating empty methods as abstract, MyPy should flag the method as missing a return statement, and optionally suggest adding the `@abstractmethod` annotation.

For example, consider this class definition:

    class P(protocol):
        def foo() -> int: ...

MyPy should raise an error like this:

    # error: Missing return statement
    # note: Did you mean to use @abstractmethod?

File: mypy/semanal.py
Comment (sobolevn) [inline mypy/semanal.py:825]: nit: I think that it is better to use `if defn.impl is None: return` here.

And remove this check: https://github.com/python/mypy/pull/12118/files#diff-f782aad4853edceb67a88efe43d2869579d075a77b8e8dd2add91e818a9fa164R816

This way we won't have accidental `AssertionError` if someone uses this method incorrectly.

**Review by sobolevn — COMMENTED** [review]
First of all, thanks for working on this!

But, sorry, I fail to understand this feature.

Questions I have:
1. I was always sure that we use protocols to define some object's *shape*, not ever use them to define *behavior*. So, I was always thinking that all protocol methods are abstract. (Moreover, I think that they should be forced to be empty, but that's a completely other discussion). Take a look at `interface` in TypeScript.
2. Why a method with `-> None` is not treated as `abstract`?
3. I think that in current state this will cause more problems that solve. For example, it would quite hard to teach which methods are abstract and which are not. I would appreciate a plan on how we can do that

**Comment by a-reich** [comment]
> First of all, thanks for working on this!
> 
> But, sorry, I fail to understand this feature.
> 
> Questions I have:
> 
> 1. I was always sure that we use protocols to define some object's _shape_, not ever use them to define _behavior_. So, I was always thinking that all protocol methods are abstract. (Moreover, I think that they should be forced to be empty, but that's a completely different discussion).

That may be how most people think of protocols, but PEP 544 chose to state that they can provide non-abstract “default implementations” and classes can rely on those. Which I think has now led to this confusion about what it means to subclass a proto and how to check it.

File: mypy/semanal.py
Comment (thomkeh) [reply mypy/semanal.py:825]: That's of course fine by me, but note that I just moved this function and didn't modify it in any way.

**Comment by thomkeh** [comment]
So, my goal was to make mypy raise an error for this case:

```python
from typing import Protocol

class Sized(Protocol):
    def __len__(self) -> int: ...

class A(Sized): ...
a = A()
assert isinstance(len(a), int)  # runtime error but no mypy error
```

The problem here is the interplay of explicit protocol-implementation (i.e., subclassing of `Sized`) and the possibility of default implementations. Without either of those things, there would be no problem. But, [the PEP](https://www.python.org/dev/peps/pep-0544/) allows for both, so here we are.

One possible solution would indeed be to force protocol methods to either provide a correct default implementation (i.e., one that returns the correct value) or to be marked as abstract:

```python
from abc import abstractmethod
from typing import Protocol

class P(Protocol):
    def meth1(self) -> int:  # E: Missing return statement
        pass

    @abstractmethod
    def meth2(self) -> int:  # OK
        pass
```
But I feel that goes a bit against what protocols were meant to do? I feel like default implementations in general are kind of not a good fit for protocols; they're better for abstract base classes, but I don't know. The advantage of this approach is that it works in stub files as well.

The other solution is to treat protocol methods that don't return what they promise to return as implicitly abstract. (Or maybe "abstract" is not the right term. Because as @nickgaya points out, protocol methods with empty default implementations aren't really "abstract" in the sense of abstract base classes; the problem is just that they implicitly return `None` (because they don't have a `return` statement) and thus should maybe simply cause a type error.) The disadvantage of this approach is that it doesn't work in stub files.

**Comment by thomkeh** [comment]
Hmm, I realize now that mypy in general doesn't complain about empty function bodies: #2350 (there was an attempt to fix it in  #8111 but it was abandoned), but I still think the problem is especially bad for protocols, so this PR should be helpful.

File: mypy/semanal.py
Comment (thomkeh) [reply mypy/semanal.py]: Sorry, I was confused about which function you were talking about. This is not the function I moved. I fixed it.

**Comment by thomkeh** [comment]
I addressed the review feedback.

There is some new relevant discussion in #8926.

Would it be better to remove the check of the return type? This would then be more in line with what pyright does.

**Comment by thomkeh** [comment]
Any interest in this? @ilevkivskyi

**Review by ilevkivskyi — COMMENTED** [review]
I like this PR. Issues with empty bodies come up quite often. About special-casing `None`: I don't think this is a good idea. In general, few people know that protocols (as well as ordinary ABCs) can have default implementations. And even among those who know and use them, few will have empty body as an _intended_ default implementation.

Also semantic analysis is a wrong place for type-checking. The current implementation already misses couple corner cases, like type variables that can be `None`, and maybe there are more.

I think it will be better to give a note that an explicit `return None` must be used if it is intentional. (IIRC we give instantiation error at type-checking stage, where you can simply use `is_subtype()`)

**Comment by thomkeh** [comment]
> About special-casing `None`: I don't think this is a good idea. In general, few people know that protocols (as well as ordinary ABCs) can have default implementations. And even among those who know and use them, few will have empty body as an _intended_ default implementation.

Yeah, that makes sense – I'll remove that special-case-ing.

> Also semantic analysis is a wrong place for type-checking. The current implementation already misses couple corner cases, like type variables that can be `None`, and maybe there are more.

I see. I will try to find the right place.

> I think it will be better to give a note that an explicit `return None` must be used if it is intentional. (IIRC we give instantiation error at type-checking stage, where you can simply use `is_subtype()`)

That sounds good. I will try to figure out how to do that.

**Comment by thomkeh** [comment]
@ilevkivskyi I implemented your suggestions (hopefully I understood them correctly), but there is a problem with `strict_optional` that I don't know how to solve. When `strict_optional` is `False`, then `is_subtype()` is basically useless when the first argument is `NoneType`. This means, it's not possible to show the helpful note in that case. I guess I could modify `is_subtype()` so that you can turn off the `strict_optional` behavior with an additional argument.

The other problem is that the `_check_non_ret_from_prot()` function is quite complicated and recomputes things that are known at other places in the code, but I think it's probably not worth caching this information from those places, because it's only needed for an error message that should be quite rare.

I'm also confused why I had to add `if not typ.is_protocol:` in `semanal_classprop.py` but without that, the stubgen tests were failing. However, it should have been previously already possible to create an abstract protocol by explicitly using `@abc.abstractmethod`.

I didn't use an `Enum` for `NOT_ABSTRACT`/`IS_ABSTRACT`/`IMPLICITLY_ABSTRACT` because to my knowledge, enums are not optimized well by mypyc, but I can change that if you want.

File: mypy/checker.py
Comment (ilevkivskyi) [inline mypy/checker.py:620]: Please don't rely on specific constants being falsey. Use explicit `==` instead. Otherwise if someone will change numeric value of the constant this code will subtly break.

File: mypy/strconv.py
Comment (ilevkivskyi) [inline mypy/strconv.py:136]: Please don't rely on specific constants being falsey. Use explicit == instead. Also please check all other places.

File: mypy/semanal.py
Comment (ilevkivskyi) [inline mypy/semanal.py:1087]: I think an overload without implementation at all should be explicitly abstract, so you should use `IS_ABSTRACT` below.

File: mypy/checkexpr.py
Comment (ilevkivskyi) [inline mypy/checkexpr.py:1385]: I think we should return False here, don't add the note if we are not sure what is the intended return type. You can simply delete the `else` branch so we fall through to last `return False`.

File: mypy/checkexpr.py
Comment (ilevkivskyi) [inline mypy/checkexpr.py:1363]: As I said above, overloads without implementation are explicitly abstract. So this whole block can be turned into `node = node.impl`

File: mypy/messages.py
Comment (ilevkivskyi) [inline mypy/messages.py:1333]: Use a simpler name for this, like `can_return_none`.

File: mypy/checkexpr.py
Comment (ilevkivskyi) [inline mypy/checkexpr.py:1348]: Give this a better name like `can_return_none_in_protocol_base`.

File: mypy/checkexpr.py
Comment (ilevkivskyi) [inline mypy/checkexpr.py:1266]: Just call this `abstract_attributes` (as the function parameter).

File: mypy/nodes.py
Comment (ilevkivskyi) [inline mypy/nodes.py:766]: Add short comment above explaining the meaning of this value.

File: mypy/nodes.py
Comment (ilevkivskyi) [inline mypy/nodes.py:2684]: Add a comment above explaining what is the meaning of this attribute (i.e. what is `int` there).

**Review by ilevkivskyi — COMMENTED** [review]
Thanks! This looks better, I have some more detailed comments now.

**Comment by ilevkivskyi** [comment]
> I'm also confused why I had to add `if not typ.is_protocol:` in `semanal_classprop.py` but without that, the stubgen tests were failing. However, it should have been previously already possible to create an abstract protocol by explicitly using `@abc.abstractmethod`.

Maybe stubgen automatically adds some imports if it sees an abstract class. In any case, I wouldn't touch `semanal_classprop` for this, fix the stubgen (or just update the test cases, if there are only few) instead.

**Comment by thomkeh** [comment]
I think I addressed everything. The code looks cleaner now, I think.

The only potential problem I see now is the `strict_optional` issue.

**Review by ilevkivskyi — APPROVED** [review]
I wouldn't worry too much about non-strict optional. Few people use it now, and it is just a usability issue, not correctness.

I have one more suggestion, after you fix it I will merge when tests pass.

File: mypy/messages.py
Comment (thomkeh) [reply mypy/messages.py]: Done. If you have suggestions on re-wording the message in any way, let me know.

**Pull request merged**
